src/nunavut/lang/py.py

```diff
@@ -0,0 +1,144 @@
+#
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# Copyright (C) 2018-2019  UAVCAN Development Team  <uavcan.org>
+# This software is distributed under the terms of the MIT License.
+#
+"""
+    Filters for generating python. All filters in this
+    module will be available in the template's global namespace as ``py``.
+"""
+import typing
+import keyword
+import builtins
+import pydsdl
+
+from .. import SupportsTemplateEnv, templateEnvironmentFilter
+from .c import VariableNameEncoder
+
+
+@templateEnvironmentFilter
+def filter_to_template_unique_name(env: SupportsTemplateEnv, base_token: str) -> str:
+    """
+    Filter that takes a base token and forms a name that is very
+    likely to be unique within the template the filter is invoked. This
+    name is also very likely to be a valid Python identifier.
+
+    .. IMPORTANT::
+
+        The exact tokens generated may change between major or minor versions
+        of this library. The only guarantee provided is that the tokens
+        will be stable for the same version of this library given the same
+        input.
+
+        Also note that name uniqueness is only likely within a given template.
+        Between templates there is no guarantee of uniqueness and,
+        since this library does not lex generated source, there is no guarantee
+        that the generated name does not conflict with a name generated by
+        another means.
+
+    Example::
+
+        {{ "foo" | py.to_template_unique_name }}
+        {{ "foo" | py.to_template_unique_name }}
+        {{ "bar" | py.to_template_unique_name }}
+        {{ "i like coffee" | py.to_template_unique_name }}
+
+    Results Example::
+
+        # These are the likely results but the specific token
+        # generated is not strongly specified.
+        _foo0_
+        _foo1_
+        _bar0_
+        _i like coffee0_ # Note that this is not a valid python identifier.
+                         # This filter does _not_ lex the base_token argument.
+
+
+    :param str base_token: A token to include in the base name.
+    :returns: A name that is likely to be valid python identifier and is likely to
+        be unique within the file generated by the current template.
+    """
+    return env.globals['_unique_name_generator']('py', base_token, '_', '_')  # type: ignore
+
+
+PYTHON_RESERVED_IDENTIFIERS = frozenset(map(str, list(keyword.kwlist) + dir(builtins)))  # type: typing.FrozenSet[str]
+
+
+def filter_id(instance: typing.Any, stropping_suffix: str = '_', encoding_prefix: str = 'ZX') -> str:
+    """
+    Filter that produces a valid Python identifier for a given object. The encoding may not
+    be reversable.
+
+    Example::
+
+        {{ "I like python" | py.id }}
+        {{ "&because" | py.id }}
+        {{ "if" | py.id }}
+        {{ "if" | py.id('stropped_') }}
+
+    Results Example::
+
+       I_like_python
+       ZX0038because
+       _if
+       stropped_if
```

Comment:
```suggestion
       ifstropped_
```
(this is probably not what you *meant* but it follows from the implementation)